### PR TITLE
GH-37532: [CI][Docs][MATLAB] Remove `GoogleTest` support from the CMake build system for the MATLAB interface

### LINKED ISSUE #37532: [CI][Docs][MATLAB] Remove `GoogleTest` support from the CMake build system for the MATLAB interface
### Describe the enhancement requested

Now, our build system has GoogleTest support but it's not used. There is only a placeholder test that tests nothing.
We had one test for `mex_util` but it removed by GH-37204.
We use the MATLAB's testing framework for all tests.

The GoogleTest support needs non-trivial CMake codes. If we can drop the GoogleTest support, we can reduce maintenance cost for it and reduce build time of the MATLAB bindings.

But we have the following use-case of the GoogleTest support:

https://github.com/apache/arrow/pull/37483#issuecomment-1702912785

> I think we want to keep the GoogleTest support for now because we do plan on adding C++ unit tests. There are a few error-conditions that cannot be triggered from MATLAB for which we want to add C++ unit tests.

Can we use another approach for the use-case?

For example, can we write a C++ method that causes an error-condition and register the method to MATLAB?

```diff
diff --git a/matlab/src/cpp/arrow/matlab/array/proxy/array.cc b/matlab/src/cpp/arrow/matlab/array/proxy/array.cc
index 9dc14c341..cedb58793 100644
--- a/matlab/src/cpp/arrow/matlab/array/proxy/array.cc
+++ b/matlab/src/cpp/arrow/matlab/array/proxy/array.cc
@@ -37,6 +37,9 @@ namespace arrow::matlab::array::proxy {
         REGISTER_METHOD(Array, type);
         REGISTER_METHOD(Array, isEqual);
 
+#ifdef ARROW_MATLAB_TEST
+        REGISTER_METHOD(Array, errorA);
+#endif
     }
 
     std::shared_ptr<arrow::Array> Array::unwrap() {
@@ -115,4 +118,11 @@ namespace arrow::matlab::array::proxy {
         mda::ArrayFactory factory;
         context.outputs[0] = factory.createScalar(is_equal);
     }
+
+#ifdef ARROW_MATLAB_TEST
+    void Array::errorA(libmexclass::proxy::method::Context& context) {
+        auto errorCondition = arrow::Status::Invalid("error A!");
+        MATLAB_ERROR_IF_NOT_OK_WITH_CONTEXT(errorCondition, context, ...ERROR_ID);
+    }
+#endif
 }
```

If we can call the method from MATLAB, can we assert it by the MATLAB testing framework?

FYI: We removed GoogleTest dependency from PyArrow by GH-14117/GH-32328 because maintaining GoogleTest support in `python/` was high cost for us.
only when 

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by kevingurney**
### Rationale for this change

This pull request removes `GoogleTest` support from the CMake build system for the MATLAB interface.

1. `GoogleTest` support adds a lot of additional complexity to the CMake build system for the MATLAB interface, and we currently don't have any standalone C++ tests for the MATLAB interface code.
2. In order to use `GoogleTest` in the MATLAB CI workflows, we are currently relying on building the tests for the Arrow C++ libraries in order to "re-use" the `GoogleTest binaries. This adds additional overhead to the MATLAB CI workflows.
3. If we want to test some internal C++ code for the MATLAB interface in the future, we can instead use a MEX function to call the code from a MATLAB test as suggested by @kou in https://github.com/apache/arrow/issues/37532#issue-1879320217.
4. There is [precedent for testing internal C++ code without GoogleTest for the Python bindings](https://github.com/apache/arrow/pull/14117).
5. On a somewhat related note - removing `GoogleTest` support will help unblock https://github.com/apache/arrow/pull/37773 as discussed in https://github.com/apache/arrow/pull/37773#issuecomment-1724207859.

### What changes are included in this PR?

1. Removed the `MATLAB_BUILD_TESTS` flag from the CMake build system for the MATLAB interface since there are no longer any C++ tests for the MATLAB interface to build.
2. Updated the `matlab_build.sh` CI workflow script to avoid building the tests for the Arrow C++ libraries and to no longer call `ctest`.
3. Updated the `README.md` for the MATLAB interface to no longer mention building or running C++ tests.
4. Updated the design document for the MATLAB Interface to no longer mention `GoogleTest` since we may end up testing internal C++ code using MEX function calls from MATLAB instead.
5. Removed placeholder C++ test (i.e. `placeholder_test.cc`).

### Are these changes tested?

Yes.

The MATLAB CI workflow is passing on all platforms.

### Are there any user-facing changes?

Yes.

There are no longer any C++ tests for the MATLAB interface. The `MATLAB_BUILD_TESTS` flag has been removed from the CMake build system to reflect this change. If a user supplies a value for `MATLAB_BUILD_TESTS` when building the MATLAB interface, the flag will be ignored by CMake.

### Future Directions

1. Add more developer-focused documentation on how to test C++ code via MEX function calls from MATLAB.

### Notes

1. In the future, we can consider testing internal C++ code using MEX function calls from MATLAB tests as suggested by @kou in https://github.com/apache/arrow/issues/37532#issue-1879320217. Currently, we don't have any C++ tests that need to be adapted to use this approach.
2. Thank you @sgilmore10 for your help with this pull request!
* Closes: #37532

### CODE REVIEW

**Comment by kevingurney** [comment]
For reference - it looks like removing `GoogleTest` support has reduced the MATLAB CI workflow by a few minutes: https://github.com/apache/arrow/actions/runs/6228155477?pr=37784.

File: matlab/doc/matlab_interface_for_apache_arrow_design.md [hunk 1]
```diff
@@ -258,7 +258,7 @@ For large tables used in a multi-process "data processing pipeline", a user coul
 ## Testing 
 To ensure code quality, we would like to include the following testing infrastructure, at a minimum: 
 1. C++ APIs 
-   - GoogleTest C++ Unit Tests 
+   - C++ Unit Tests
```
Comment (kou) [inline matlab/doc/matlab_interface_for_apache_arrow_design.md:261]: Can we remove this item?

**Review by kou — APPROVED** [review]
+1

**Comment by kou** [comment]
CI time on main: https://github.com/apache/arrow/actions/runs/6224657359

* Ubuntu: 19m 41s
* macOS: 24m 47s
* Windows: 33m 0s

CI time on this PR: https://github.com/mathworks/arrow/actions/runs/6228155036

* Ubuntu: 7m 53s
* macOS: 11m 3s
* Windows: 10m 43s

**Review by assignUser — APPROVED** [review]
Looks good, always a fan of reducing complexity :D 

One thing for a follow up would be to use fetchContent instead of externalProject to build libarrow, that way you don't have to manually create the targets and add the imported location etc.

**Comment by assignUser** [comment]
Ah that's true... we should fix that ^^

Though @amoeba has an [example](https://github.com/amoeba/arrow-cmake-fetchcontent) but that needs a hack so I am inclined to agree the ep is the better choice for now.

**Comment by kevingurney** [comment]
> we should fix that ^^

Agreed! Supporting `FetchContent` is definitely something that seems like it could help simplify adoption of the Arrow C++ libraries in other projects.

> Though @amoeba has an [example](https://github.com/amoeba/arrow-cmake-fetchcontent)

Thanks for sharing this! I wasn't aware there was a working example.

As you said, I think that sticking with `ExternalProject` for the time being is a good idea, given the need for the hack. However, we would be happy to move over to `FetchContent` if this gets resolved in the future.

File: matlab/doc/matlab_interface_for_apache_arrow_design.md (hunk 1 above)
Comment (kevingurney) [reply matlab/doc/matlab_interface_for_apache_arrow_design.md:261]: That's a good question.

I think it's fine to remove this for the time being (or, at least reword this section to talk about the MEX-based C++ testing approach).

The thinking was that we might want to add some C++ tests in the future. However, we haven't yet had a compelling need to do that, so I think it makes sense to change the wording here.

Comment (kevingurney) [reply matlab/doc/matlab_interface_for_apache_arrow_design.md]: I restructured the testing section slightly and added a note about testing internal C++ code via MEX function calls from a MATLAB Class-Based Unit test.

Comment (kevingurney) [reply matlab/doc/matlab_interface_for_apache_arrow_design.md]: Marking this as resolved.

**Comment by kevingurney** [comment]
+1

**Pull request merged**